### PR TITLE
[Refactor] Refactor in subquery transform rule

### LINKED ISSUE #5856: query result error
### Steps to reproduce the behavior (Required)
```
DROP DATABASE IF EXISTS w19;
CREATE DATABASE w19;
USE w19;
admin set frontend config ("enable_decimal_v3" = "true");
CREATE TABLE t0 (c_0_0 DATE NOT NULL ,c_0_1 DATE NOT NULL ,c_0_2 CHAR(51) NULL ,c_0_3 BOOLEAN NOT NULL ,c_0_4 VARCHAR(21) NULL ,c_0_5 DATE NOT NULL ) UNIQUE KEY (c_0_0,c_0_1,c_0_2,c_0_3) PARTITION BY RANGE(c_0_1) (START ("2010-01-01") END ("2021-12-31") EVERY (INTERVAL 2191 day)) DISTRIBUTED BY HASH (c_0_0,c_0_1,c_0_2,c_0_3) BUCKETS 3 properties("replication_num"="3") ;
CREATE TABLE t1 (c_1_0 DATE NULL ,c_1_1 DATE NOT NULL ,c_1_2 VARCHAR(1) NOT NULL ,c_1_3 DECIMAL(13, 8)  NULL ,c_1_4 HLL HLL_UNION NOT NULL ,c_1_5 DECIMAL(1, 0)  REPLACE NULL ,c_1_6 DOUBLE SUM NULL ,c_1_7 DOUBLE MIN NOT NULL ,c_1_8 CHAR(31) REPLACE NOT NULL ,c_1_9 PERCENTILE PERCENTILE_UNION NULL ,c_1_10 PERCENTILE PERCENTILE_UNION NOT NULL ,c_1_11 BOOLEAN REPLACE NULL ,c_1_12 PERCENTILE PERCENTILE_UNION NOT NULL ,c_1_13 HLL HLL_UNION NULL ,c_1_14 VARCHAR(11) REPLACE NOT NULL ,c_1_15 DECIMAL(12, 1)  SUM NULL ,c_1_16 DECIMAL(35, 10)  MIN NULL ,c_1_17 HLL HLL_UNION NULL ) AGGREGATE KEY (c_1_0,c_1_1,c_1_2,c_1_3) PARTITION BY RANGE(c_1_0) (START ("2010-01-01") END ("2021-12-31") EVERY (INTERVAL 101 day)) DISTRIBUTED BY HASH (c_1_3,c_1_1,c_1_2,c_1_0) BUCKETS 3 properties("replication_num"="3") ;
CREATE TABLE t2 (c_2_0 INT NULL ,c_2_1 DATE NULL ,c_2_2 FLOAT SUM NULL ,c_2_3 DECIMAL(26, 17)  REPLACE NULL ,c_2_4 CHAR(31) MAX NULL ,c_2_5 HLL HLL_UNION NOT NULL ,c_2_6 DECIMAL(14, 12)  MIN NOT NULL ,c_2_7 DATE MAX NOT NULL ,c_2_8 FLOAT REPLACE NOT NULL ,c_2_9 DATE MAX NULL ,c_2_10 CHAR(11) MAX NULL ) AGGREGATE KEY (c_2_0,c_2_1) DISTRIBUTED BY HASH (c_2_0,c_2_1) BUCKETS 3 properties("replication_num"="3") ;
CREATE TABLE t3 (c_3_0 DATETIME NOT NULL ,c_3_1 VARCHAR(21) NULL ,c_3_2 DATETIME NULL ,c_3_3 JSON NULL ,c_3_4 ARRAY<INT> NOT NULL ,c_3_5 DECIMAL(14, 1)  NULL ,c_3_6 DOUBLE NULL ,c_3_7 DOUBLE NULL ,c_3_8 DECIMAL(5, 1)  NULL ,c_3_9 BOOLEAN NULL ,c_3_10 FLOAT NOT NULL ,c_3_11 DOUBLE NULL , INDEX index1_c_3_0(`c_3_0`) USING BITMAP) DUPLICATE KEY (c_3_0,c_3_1,c_3_2) PARTITION BY RANGE(c_3_0) (START ("2010-01-01") END ("2021-12-31") EVERY (INTERVAL 93 day)) DISTRIBUTED BY HASH (c_3_2,c_3_1,c_3_0) BUCKETS 3 properties("replication_num"="3") ;
CREATE TABLE t4 (c_4_0 DATETIME NOT NULL ,c_4_1 BOOLEAN NULL ,c_4_2 BOOLEAN NOT NULL ,c_4_3 DATE NOT NULL ,c_4_4 DATETIME NULL ,c_4_5 LARGEINT NOT NULL ,c_4_6 DECIMAL(31, 9)  NOT NULL ,c_4_7 INT NOT NULL ,c_4_8 DECIMAL(26, 19)  NULL ,c_4_9 DATETIME NOT NULL ,c_4_10 DATE NULL ,c_4_11 CHAR(21) NULL ,c_4_12 BIGINT NULL ,c_4_13 DECIMAL(27, 5)  NULL ,c_4_14 DATETIME NOT NULL ,c_4_15 DATE NULL ,c_4_16 DOUBLE REPLACE_IF_NOT_NULL NULL ) AGGREGATE KEY (c_4_0,c_4_1,c_4_2,c_4_3,c_4_4,c_4_5,c_4_6,c_4_7,c_4_8,c_4_9,c_4_10,c_4_11,c_4_12,c_4_13,c_4_14,c_4_15) PARTITION BY RANGE(c_4_0) (START ("2010-01-01") END ("2021-12-31") EVERY (INTERVAL 87 day)) DISTRIBUTED BY HASH (c_4_7,c_4_1,c_4_9,c_4_14,c_4_3,c_4_0,c_4_6,c_4_11,c_4_4,c_4_10) BUCKETS 3 properties("replication_num"="3", "bloom_filter_columns"="c_4_7") ;
DELETE FROM t3 WHERE c_3_1="bkITS";
INSERT INTO t3 VALUES ("2010-01-09 08:00:14", "t", "2010-01-08 08:00:14", parse_json('{"D": null, "OAF": 0.4218158163647504, "9": 0.7525813488619105}'), [1049894433, 2095209568, 1049894433, -1974592483, -141277720, -2020911495, -141277720, -979704951, -2020911495, -2020911495, -2020911495, 503922890, -1899295097, 1435760539], 934573990883.8, 0.6653126833095009, 0.7575358224102343, 711.1, false, 1.435760539E9, 0.6653126833095009), ("2010-01-22 08:00:14", "t", "2010-01-10 08:00:14", parse_json('{"44": false, "Bl2Mx": 0.7575358224102343, "d": "", "e": null, "JFje0": "N", "n": null, "N": true, "ZI": false, "yatxFRu": null, "Rdrcc": -1875888215, "RwIO1AOMC": null, "iHhgsmg": 0.06349533566736443, "LaZd": "ekKZTWFg", "jE2D": null}'), [1090166231, -1413815547, 1436294532, -1474040615, -1412617045, 1607764445, 1703990495], 206294329886.0, -2.020911495E9, 0.25823589645464595, 928.1, false, '+Inf', 0.2780801213352383);
INSERT INTO t2 VALUES (5, "2010-01-01", 0.2595016243340347, 5.18165636E-9, "", hll_hash("6FL"), 8.6E-11, "2010-01-28", -1.875888215E9, "2010-01-19", "qj"), (NULL, "2010-01-03", 0.7525813488619105, NULL, "", hll_hash("8oONvV1VY"), 2.5E-11, "2010-01-22", -2.020911495E9, "2010-01-01", "");
INSERT INTO t3 VALUES ("2009-12-31 08:00:14", "rU", "2010-01-29 08:00:14", parse_json('{"g": [false, null, [null, true, true, false, null, null, null, true], "", {"Wnx": false, "A": 503922890, "vVWtoBmfH": "FK", "Ve38dow": null, "acmbd0a": null, "fLYY": true, "M": null, "LDUpr": null, "S": 0.8509869792167217, "W": null, "su7TPp": false, "y": "X5Pqs", "Y6": true, "F9R": 1590709569, "Dv099wv": -2010715141}, "5", {"Y6lzJam": "hq", "WsE": null, "3qUDCQOp": "DeKz3a", "QWTx": null, "Xw": true}, false, {"vv": null, "f": false, "O3j58AuG": 0.7157648475406833, "7jlNe9": "UKl9hCC", "wg": 1110877422, "lEDrN3DUp": "b05YN7aM", "H9": true, "RdFmMw": 0.20135271189824921, "t": false, "Mb": null, "V": null, "y": null, "u1": 0.6476661452665934}, "H", [true, 0.367823379991739, "", 0.8773548510806943], 0.09106685260780556, ""], "gMo": [-2020911495, true, {"Uv": 5.0392289E8, "J4uIK": null, "gyw": true, "M": 0.4044190487426865, "N": -2010715141, "I8": true, "2": "53xTkV9A9", "eJsrY89n2": null, "6": 5.0392289E8, "sl": -340518995, "MTxxTBSU": false, "e0uOKPiX": -1875888215, "So": 0.2780801213352383, "T36na": "aJF", "ur": true}, -454203382, true, 0.40372657774621634, -141277720, -9223372036854775808, null, ["wCGlnd", true, "Prc", 0.6653126833095009, 0.3831336488506494, "Nf3", false, false, ""], null, false, 1059720520, true], "wLzzWO9": -1.4127772E8, "19Noy74": -2020911495, "tUt4jL88w": {"2": "Vfz9q", "bxP4vux": "", "RZwfW": {"A": null, "Hj": false, "N1f": 195811771, "T": "tv", "GUg0": true, "a4sk": true, "O7": -1386569082, "i": "4", "9": 0.9075550106822707, "3ya": -0.0, "j8H": 0.06349533566736443, "YamVH4kG": null}}, "hoC": 1049894433, "yv": [0.3831336488506494, 1778370041, false, {"Q": 0.4470770184969738, "Zl": null, "p3": null, "nQ": true, "U": 0.6593125576661847, "v": 41037987, "GT7mBHxex": false, "0bQNKd": null, "yoKujsG": true, "fPKotOgw": null, "N": 0.4081581475503403}, "1mQoCivv", false, ["u8R", 1049894433]], "n": ""}'), [250227185, -282187627, -111534546, -121267743, 520834961, 1545903730, 1553886427, 1562340703, 895820061, -340518995, 1969302940], 126402464013.9, 0.7657969101237654, 1.54590373E9, NULL, false, 0.5312557739617265, 0.5194464221478821), ("2010-01-01 08:00:14", NULL, "2010-01-31 08:00:14", parse_json('{"S": "", "4": {"R21": -1410637030, "3": null, "4h": false, "oiRz": 1435760539}, "aZeHHy": 0.016135975667814884, "7": 2069012194, "niDwQfg": {"mLA7IMy5": "VhveQhq2r"}, "DP": null, "TROJd": {"Z5TeKvg1y": 0.4044190487426865, "1yWTEh": 371121197, "7": 0.2181739230916544, "e2JHkCI": false}, "03QY": null, "gAGj2": -977331514, "DC": null}'), [520834961, -1474040615, 1607764445, -1516095629, -651331120, -666222901, -27927492, -194224740], 534564424139.8, 0.2606909442679777, 0.6115737711099407, 381.6, true, 0.9773811994709547, -1.9422474E8);
INSERT INTO t4 VALUES ("2010-01-20 08:00:14", true, true, "2009-12-31", "2010-01-21 08:00:14", 1562340703, 8499831709653.454914555, 8, 2.208558E-13, "2010-01-15 08:00:14", "2010-01-30", "UQYcO", -121267743, 72741205874020750.48776, "2009-12-31 08:00:14", "2010-01-02", 3.71121197E8), ("2010-01-14 08:00:14", false, false, "2010-01-22", "2010-01-17 08:00:14", -528934991, 6435686077969.174033001, 23, 8.913380E-13, "2010-01-30 08:00:14", "2010-01-15", NULL, 1553886427, 71104956802292925.02379, "2010-01-01 08:00:14", "2010-01-20", -1.474040615E9);
INSERT INTO t1 VALUES ("2010-01-03", "2010-01-29", "", 0.00045819, hll_hash("kSv9im"), 5, 0.31916937249764454, 0.4218158163647504, "td", percentile_hash("1200609970"), percentile_hash("503922890"), true, percentile_hash("-345510899"), hll_hash("YQ"), "3", 5119085577.1, 20256295393006.7735087463, hll_hash("nmL3t")), ("2010-01-18", "2010-01-20", "", 0.00071961, hll_hash("xo6O"), 5, -1.4127772E8, 1.96930294E9, "", percentile_hash("1590709569"), percentile_hash("1436294532"), NULL, percentile_hash("1382130056"), hll_hash("hFCXl"), "Y", 6396243630.8, 920450594242091.7463665284, hll_hash("7aaE")), ("2010-01-10", "2010-01-02", "", 0.00062116, hll_hash("kl"), 2, 0.027547055351589744, 0.6906040762443255, "GPUJy", percentile_hash("-1272725795"), percentile_hash("-403446248"), false, percentile_hash("-253510490"), hll_hash("CMe0JknI"), "JwXvce", 5680451054.8, 815778667420325.6538362438, hll_hash("a")), ("2010-01-23", "2010-01-06", "", 0.00073043, hll_hash("c"), 2, 0.5261522689793094, 0.7657969101237654, "1", percentile_hash("-1860101835"), percentile_hash("-528934991"), false, percentile_hash("371121197"), hll_hash("EYo94HY"), "", 3582895901.4, 102323327441853.2985805541, hll_hash("BfvJjh"));
INSERT INTO t1 VALUES ("2010-01-23", "2010-01-07", "", 0.00013191, hll_hash(""), 7, 0.49744876459617715, -3.40518995E8, "aoQ3weC8", percentile_hash("964512680"), percentile_hash("-588212140"), true, percentile_hash("-141277720"), hll_hash(""), "6", 4038790518.9, 200906373040653.2854999936, hll_hash("HY60Z")), ("2010-01-05", "2010-01-28", "", 0.00043585, hll_hash(""), 7, NULL, 0.5465855181645083, "pt", percentile_hash("-141277720"), percentile_hash("-1793536092"), false, percentile_hash("1312687050"), hll_hash("r"), "rr2", 2180322047.7, 137081916397446.1574184743, hll_hash("At"));
INSERT INTO t4 VALUES ("2010-01-24 08:00:14", NULL, false, "2010-01-08", "2010-01-18 08:00:14", 473073076, 588723143143.363294862, 19, 7.187194E-13, "2010-01-01 08:00:14", "2010-01-26", "Z", 1059720520, 73476688045904527.30526, "2010-01-08 08:00:14", "2010-01-26", 0.38202206908938263);
INSERT INTO t2 VALUES (14, "2010-01-14", 1.20060997E9, NULL, "1qKaArI", hll_hash("x0HjZs"), 3.7E-11, "2010-01-12", 0.6001685470349415, "2010-01-05", NULL), (8, "2010-01-02", 0.9151045173539579, 7.48060634E-9, "", hll_hash("V9QyLv"), 4.2E-11, "2010-01-28", 0.9555137711637539, "2010-01-02", "DHC96");
INSERT INTO t2 VALUES (28, NULL, 0.11087009306956075, 1.66074155E-9, "", hll_hash("Wma"), 7.7E-11, "2010-01-02", 0.5730863836497143, "2010-01-18", "");

SELECT t1_187.c_1_0 FROM t1 AS t1_187;
-- cardinality: 6
SELECT t1_187.c_1_0 FROM t1 AS t1_187 WHERE (t1_187.c_1_6) IN ((SELECT t1_187.c_1_6 FROM t1 AS t1_187 WHERE true) )  UNION ALL SELECT t1_187.c_1_0 FROM t1 AS t1_187 WHERE (NOT ((t1_187.c_1_6) IN ((SELECT t1_187.c_1_6 FROM t1 AS t1_187 WHERE true) ) )) UNION ALL SELECT t1_187.c_1_0 FROM t1 AS t1_187 WHERE (((t1_187.c_1_6) IN ((SELECT t1_187.c_1_6 FROM t1 AS t1_187 WHERE true) ) ) IS NULL);
-- cardinality: 5

```

### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
 - mysql> select current_version();
+----------------------------+
| current_version()          |
+----------------------------+
| BRANCH-2.2-RELEASE 00a537f |
+----------------------------+
- cluster: sqlancer_branch_release

### PR DESCRIPTION
**Pull request by Seaven**
## What type of PR is this：
- [x] bug
- [ ] feature
- [ ] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #5856 #5857 #5955

## Problem Summary(Required) ：

SQL:
```
select t0.v1, t0.v2 in (select t1.v2 from t1)
from t0;
```

will transform to sql now:
```
select t0.v1, x1.v2 is not null 
from t0 left outer join (select distinct v2 from t1) as x1 on t0.v2 = x1.v2
```

the SQL will ignore NULL resolve, the result as:

IN | result| after transform, the t1.v2 | t1.v2 is not null
-- | -- | -- | --
1 in (1) | 1 | 1 | 1
1 in (2) | 0 | NULL | 0
1 in (2, null) | NULL | NULL | 0
1 in (1, 2, null) | 1 | 1 | 1
Null in (1, 2) | NULL | NULL | 0
Null in (1, null) | NULL | NULL | 0

And refactor in subquery transform rule.
```
        /*
         * In:
         * before: select t0.v1, t0.v2 in (select t1.v2 from t1 where t0.v3 = t1.v3) from t0;
         * after: with xx as (select t1.v2, t1.v3 from t1)
         *        select t0.v1,
         *             case
         *                 when t1Rows = 0 or t1Rows is null then false // t1 empty table
         *                 when t0.v2 is null then null
         *                 when t1d.v2 is not null then true
         *                 when v2Nulls < t1Rows then null
         *                 else false
         *             end
         *         from t0
         *              left outer join (select xx.v2, xx.v3 from xx group by xx.v2, xx.v3) as t1d
         *                              on t0.v2= t1d.v2 and t0.v3 = t1d.v3
         *              left outer join (select count(*) as t1Rows, count(xx.v2) as v2Nulls from xx group by xx.v3) as t1c
         *                              on t1c.v3 = t0.v3
         *
         *                                           CTEAnchor
         *                                          /        \
         *                                    CTEProduce     Project
         *    Apply(t0.v2 in t1.v2)              /               \
         *    /          \          ===>     Project         left/cross join
         *   t0          t1                   /              /          \
         *                                Filter         Left join       Agg(count)
         *                                 /             /       \           \
         *                               t1            t0     Agg(distinct)  CTEConsume
         *                                                         \
         *                                                        CTEConsume
         * */
```

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/ReplaceColumnRefRewriter.java [hunk 1]
```diff
@@ -28,6 +28,9 @@ public ReplaceColumnRefRewriter(Map<ColumnRefOperator, ScalarOperator> operatorM
     }
 
     public ScalarOperator rewrite(ScalarOperator origin) {
+        if (origin == null) {
```
[the remark this answers is not in the document]
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/ReplaceColumnRefRewriter.java:31]: Cross join's predicate

**Comment by Seaven** [comment]
run starrocks_fe_unittest

**Comment by Youngwb** [comment]
t1Rows = 0 and t0.v2 is null return false?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/QuantifiedApply2OuterJoinRule.java [hunk 2]
```diff
@@ -29,22 +55,346 @@ public boolean check(OptExpression input, OptimizerContext context) {
...
 
-        return transform2OuterJoin(context, input, apply, correlationPredicates, correlationColumnRefs, ipo.isNotIn());
+        // check correlation filter
+        if (!SubqueryUtils.checkAllIsBinaryEQ(joinOnPredicate, correlationColumnRefs)) {
+            // @Todo:
+            // 1. require least a EQ predicate
+            // 2. for other binary predicate rewrite rule
+            //  a. outer-key < inner key -> outer-key < aggregate MAX(key)
+            //  b. outer-key > inner key -> outer-key > aggregate MIN(key)
+            throw new SemanticException("Not support Non-EQ correlation predicate correlation in sub-query");
```
[the remark this answers is not in the document]
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/QuantifiedApply2OuterJoinRule.java:99]: Must check after finished decorrelate.

**Comment by Seaven** [comment]
> t1Rows = 0 and t0.v2 is null return false?

t1 is empty table or  t0.v2 is null

**Comment by Seaven** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 176 / 196 (89.80%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/ExistentialApply2OuterJoinRule.java|31|40|77.50%|
|:large_blue_circle:|com/starrocks/sql/optimizer/SubqueryUtils.java|11|13|84.62%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/QuantifiedApply2OuterJoinRule.java|115|124|92.74%|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|3|3|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownApplyProjectRule.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalCTEAnchorOperator.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/OperatorBuilderFactory.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/CTEAnchorImplementationRule.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalCTEAnchorOperator.java|8|8|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/ReplaceColumnRefRewriter.java|2|2|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 176 / 196 (89.80%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/ExistentialApply2OuterJoinRule.java|31|40|77.50%|
|:large_blue_circle:|com/starrocks/sql/optimizer/SubqueryUtils.java|11|13|84.62%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/QuantifiedApply2OuterJoinRule.java|115|124|92.74%|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|3|3|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownApplyProjectRule.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalCTEAnchorOperator.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/OperatorBuilderFactory.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/CTEAnchorImplementationRule.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalCTEAnchorOperator.java|8|8|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/ReplaceColumnRefRewriter.java|2|2|100.00%|

**Comment by Seaven** [comment]
run starrocks_fe_unittest

**Review by Youngwb — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownApplyProjectRule.java [hunk 3]
```diff
@@ -57,7 +57,7 @@ public List<OptExpression> transform(OptExpression input, OptimizerContext conte
         newApply.getInputs().addAll(child.getInputs());
 
         ColumnRefFactory factory = context.getColumnRefFactory();
-        Map<ColumnRefOperator, ScalarOperator> allOutput = Maps.newHashMap(project.getColumnRefMap());
```
Comment (HangyuanLiu) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownApplyProjectRule.java:60]: why we can delete project

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/QuantifiedApply2OuterJoinRule.java [hunk 4]
```diff
@@ -29,22 +55,361 @@ public boolean check(OptExpression input, OptimizerContext context) {
...
+            countNulls = factory.create("countNulls", countNullFn.getType(), countNullFn.isNullable());
+            aggregates.put(countNulls, countNullFn);
+
+            return OptExpression.create(
+                    new LogicalAggregationOperator(AggType.GLOBAL, countAggregateGroupBys, aggregates), countConsume);
+        }
+
+        /*
+         * 1. build cross join if un-correlation sub-query
+         * 2. build left join if correlation sub-query
```
Comment (HangyuanLiu) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/QuantifiedApply2OuterJoinRule.java:360]: Please add a note explaining why there is a difference between cross and left outer

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/logical/LogicalApplyOperator.java [hunk 5]
```diff
@@ -112,7 +112,6 @@ public boolean isUseSemiAnti() {
     public ColumnRefSet getOutputColumns(ExpressionContext expressionContext) {
         ColumnRefSet outputColumns =
                 (ColumnRefSet) expressionContext.getChildLogicalProperty(0).getOutputColumns().clone();
-        outputColumns.union(expressionContext.getChildLogicalProperty(1).getOutputColumns());
```
[the remark this answers is not in the document]
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/logical/LogicalApplyOperator.java:115]: It's error, subquery can't output inner table columns

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownApplyProjectRule.java (hunk 3 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownApplyProjectRule.java:60]: Subquery can't output inner table columns

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 179 / 199 (89.95%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/ExistentialApply2OuterJoinRule.java|31|40|77.50%|
|:large_blue_circle:|com/starrocks/sql/optimizer/SubqueryUtils.java|11|13|84.62%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/QuantifiedApply2OuterJoinRule.java|118|127|92.91%|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|3|3|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownApplyProjectRule.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalCTEAnchorOperator.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/OperatorBuilderFactory.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/CTEAnchorImplementationRule.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalCTEAnchorOperator.java|8|8|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/ReplaceColumnRefRewriter.java|2|2|100.00%|

**Review by Youngwb — APPROVED** [review]

**Review by HangyuanLiu — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/QuantifiedApply2OuterJoinRule.java [hunk 6]
```diff
@@ -29,22 +55,362 @@ public boolean check(OptExpression input, OptimizerContext context) {
                 && !Utils.containsCorrelationSubquery(input.getGroupExpression());
     }
 
+    /*
+     *
+     * @todo:
```
[the remark this answers is not in the document]
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/QuantifiedApply2OuterJoinRule.java:60]: I think we need support the case later

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/QuantifiedApply2OuterJoinRule.java [hunk 7]
```diff
@@ -29,22 +55,362 @@ public boolean check(OptExpression input, OptimizerContext context) {
...
+         *                 when t1Rows = 0 or t1Rows is null then false // t1 empty table
+         *                 when t0.v2 is null then null
+         *                 when t1d.v2 is not null then true
+         *                 when v2Nulls < t1Rows then null
+         *                 else false
+         *             end
+         *         from t0
+         *              left outer join (select xx.v2, xx.v3 from xx group by xx.v2, xx.v3) as t1d
+         *                              on t0.v2= t1d.v2 and t0.v3 = t1d.v3
+         *              left outer join (select count(*) as t1Rows, count(xx.v2) as v2Nulls from xx group by xx.v3) as t1c
```
[the remark this answers is not in the document]
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/QuantifiedApply2OuterJoinRule.java:169]: Yeah, you are right

**Review by liuyehcf — APPROVED** [review]

**Pull request merged**